Sources/Algorithms/Keyed.swift

```diff
@@ -0,0 +1,86 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift Algorithms open source project
+//
+// Copyright (c) 2020 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+//
+//===----------------------------------------------------------------------===//
+
+public struct KeysAreNotUnique<Key, Element>: Error {
+  public let key: Key
+  public let previousElement: Element
+  public let conflictedElement: Element
+
+  public init(key: Key, previousElement: Element, conflictedElement: Element) {
+    self.key = key
+    self.previousElement = previousElement
+    self.conflictedElement = conflictedElement
+  }
+}
+
+extension Sequence {
+  /// Creates a new Dictionary from the elements of `self`, keyed by the
+  /// results returned by the given `keyForValue` closure. Deriving the
+  /// same duplicate key for more than one element of `self` will cause
+  /// an error to be thrown.
+  ///
+  /// - Parameters:
+  ///   - keyForValue: A closure that returns a key for each element in `self`.
+  @inlinable
+  public func keyed<Key>(
+    by keyForValue: (Element) throws -> Key
+  ) throws -> [Key: Element] {
+    var result = [Key: Element]()
+
+    for element in self {
+      let key = try keyForValue(element)
+
+      if let previousElement = result.updateValue(element, forKey: key) {
+        throw KeysAreNotUnique(key: key, previousElement: previousElement, conflictedElement: element)
+      }
+    }
+
+    return result
```

Comment:
Would it make sense to have this in terms of the other overload? e.g.
```suggestion
    return try self.keyed(by: keyForValue, resolvingConflictsWith: { key, previousElt, conflictedElt in 
       throw KeysAreNotUnique(key: key, previousElement: previousElement, conflictedElement: element)
    })
```

Reply:
Oooo fantastic! This was a case of path-dependence, it wasn't possible before, but now that it's two functions, it's available and much nicer!